src/client.ts

```diff
@@ -5234,15 +5234,15 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
      * <p>If the EventTimelineSet object already has the given event in its store, the
      * corresponding timeline will be returned. Otherwise, a /context request is
      * made, and used to construct an EventTimeline.
-     * If the event does not belong to this EventTimelineSet then undefined will be returned.
+     * If the event does not belong to this EventTimelineSet then it will ignored.
```

Comment:
The event will be ignored? What does that even mean? This method isn't meant to process the given event, just use it as a pointer. What event timeline will be returned to the caller? The caller would now have need to manually check that the returned timeline is valid for what they asked for, I guess by your other change lower down by attempting to add an event and by asserting that it worked, that seems rather strange

Reply:
Let's first discuss the previous behavior. Previously, when the event didn't belong to the `timelineSet`, we would return `undefined` and create no additional timelines. Returning `undefined` has no special meaning given that all of the usage is always fire and forget meaning we don't use the `timeline` returned by `client.getEventTimeline()` and only use the function to load the event so it's available to the client.

In our usage, there is only a single spot where the caller [uses the `timeline` returned by `client.getEventTimeline()`](https://github.com/matrix-org/matrix-react-sdk/blob/f5f79158ed5a950626bc89439b8319e362a75de1/src/components/structures/LoggedInView.tsx#L356). This usage should be replaced by `room.findEventById(eventId)` because it doesn't even use the `timeline`, it just wants the event that was loaded in as well.

It's probably a misnomer to call it `getEventTimeline(): timeline` in the first place as it's more accurately used as `loadEventInTimeline(): Promise<void>` everywhere.

---

We're only working within the given `timelineSet` that was passed in. If `client.getEventTimeline()` is really meant to just give the timeline for the `eventId`, then we should just provide the `room` instead which can look at all of the timelineSets (`room.timelineSets`).

Previously, we would return `undefined` in the case where the event doesn't belong in the `timelineSet`. Now we're returning a `timeline` in the `timelineSet` where all of the events returned by `/context` can go. This means events that we fetched, are actually added and not wasted. And it means that the main room timeline can be populated regardless if the `eventId` passed in was a threaded reply.

By ignored, I mean if the `eventId` is a threaded reply, it won't be added to the main room `timelineSet` for example.

Reply:
Our usage might be to be a fire-and-forget pattern (or wrong, in the case of needing to use `findEventById` instead), but as a public function and SDK we have to maintain a rationalized contract for the function name: it says it gets an event timeline, so it should do that (returning `undefined` if needed)

We can adjust our code to instead use a new `updateEventTimeline()` function or similar, but the existing `getEventTimeline` function can't realistically have a behavioural change like this.

Reply:
The `return undefined` part is part of `supportsExperimentalThreads` which changed in https://github.com/matrix-org/matrix-js-sdk/pull/2444/files

Can we change the experimental implementation?

Reply:
I'm still not comfortable with the behavioural change here, sorry. While our usage might be fire-and-forget, we can't guarantee that *all* usages of the function are fire-and-forget. The documentation and function itself are not experimental in nature as well, preventing us from making arbitrary breaking changes.

Adding a function is more code, but I think it's worthwhile here. It can even call `getEventTimeline()` and ignore the return value - it looks a bit silly, but it's how we avoid unnecessary major version releases.

Reply:
Conversation continued at https://github.com/matrix-org/matrix-js-sdk/pull/2852#discussion_r1013894123